Signal/src/ViewControllers/HomeView/HomeViewCell.m

```diff
@@ -0,0 +1,424 @@
+//
+
+#import "HomeViewCell.h"
+#import "OWSAvatarBuilder.h"
+#import "Signal-Swift.h"
+#import <SignalMessaging/OWSFormat.h>
+#import <SignalMessaging/OWSUserProfile.h>
+#import <SignalMessaging/SignalMessaging-Swift.h>
+#import <SignalServiceKit/OWSMessageManager.h>
+#import <SignalServiceKit/TSContactThread.h>
+#import <SignalServiceKit/TSGroupThread.h>
+#import <SignalServiceKit/TSThread.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+const NSUInteger kHomeViewCellHeight = 72;
+const NSUInteger kHomeViewCellHMargin = 16;
+const NSUInteger kHomeViewCellVMargin = 12;
+const NSUInteger kHomeViewAvatarSize = kHomeViewCellHeight - kHomeViewCellVMargin * 2;
+const NSUInteger kHomeViewAvatarHSpacing = 12;
+
+@interface HomeViewCell ()
+
+@property (nonatomic) AvatarImageView *avatarView;
+@property (nonatomic) UIView *payloadView;
+@property (nonatomic) UILabel *nameLabel;
+@property (nonatomic) UILabel *snippetLabel;
+@property (nonatomic) UILabel *dateTimeLabel;
+@property (nonatomic) UIView *unreadBadge;
+@property (nonatomic) UILabel *unreadLabel;
+
+@property (nonatomic, nullable) TSThread *thread;
+@property (nonatomic, nullable) OWSContactsManager *contactsManager;
+
+@property (nonatomic, readonly) NSMutableArray<NSLayoutConstraint *> *viewConstraints;
+
+@end
+
+#pragma mark -
+
+@implementation HomeViewCell
+
+- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(nullable NSString *)reuseIdentifier
+{
+    if (self = [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
+        [self commontInit];
+    }
+    return self;
+}
+
+// `[UIView init]` invokes `[self initWithFrame:...]`.
+- (instancetype)initWithFrame:(CGRect)frame
+{
+    if (self = [super initWithFrame:frame]) {
+        [self commontInit];
+    }
+
+    return self;
+}
+
+- (void)commontInit
+{
+    OWSAssert(!self.avatarView);
+
+    [self setTranslatesAutoresizingMaskIntoConstraints:NO];
+    self.layoutMargins = UIEdgeInsetsZero;
+    self.contentView.layoutMargins = UIEdgeInsetsZero;
+    self.preservesSuperviewLayoutMargins = NO;
+    self.contentView.preservesSuperviewLayoutMargins = NO;
+
+    self.backgroundColor = [UIColor whiteColor];
+
+    _viewConstraints = [NSMutableArray new];
+
+    self.avatarView = [[AvatarImageView alloc] init];
+    [self.contentView addSubview:self.avatarView];
+    [self.avatarView autoSetDimension:ALDimensionWidth toSize:kHomeViewAvatarSize];
+    [self.avatarView autoSetDimension:ALDimensionHeight toSize:kHomeViewAvatarSize];
+    [self.avatarView autoPinLeadingToSuperviewMarginWithInset:kHomeViewCellHMargin];
+    [self.avatarView autoVCenterInSuperview];
+    [self.avatarView setContentHuggingHigh];
+    [self.avatarView setCompressionResistanceHigh];
+
+    self.payloadView = [UIView containerView];
+    [self.contentView addSubview:self.payloadView];
+    [self.payloadView autoPinLeadingToTrailingEdgeOfView:self.avatarView offset:kHomeViewAvatarHSpacing];
+    [self.payloadView autoVCenterInSuperview];
```

Comment:
Seems like it would be better to pin to superview height, rather than vcenter, so as not to grow beyond it's bounds, or maybe it could be a vertical stack view and use the "spacing" attribute to adjust the gap between.

Also, if payload itself were a stack view, we would have access to some maybe useful sizing behavior that might allow us to adapt to dynamic text more easily.

Just food for thought, not sure if that would actually make any of this easier.

Reply:
I'm going to address this in the next PR to avoid rebase hell.